### PR TITLE
Model

### PR DESCRIPTION
**Pull request by swabhs**
This PR contains:

1. Segmental Language Model (no more dependence on Calypso)
2. Tests and fixtures for Segmental Language Model
3. A perplexity metric
4. Altering the BiLMTransformer module to incorporate directions (for segmental layers).

### CODE REVIEW

File: allennlp/data/dataset_readers/segmental_conll2000.py [hunk 1]
```diff
@@ -164,8 +164,8 @@ def text_to_instance(self, # type: ignore
         # However, each O span needs to be of length 1, since there is no reason to
         # combine tokens with O tags as a span, hence replacing O with U-O.
         chunk_tags = ['U-O' if tag == 'O' else tag for tag in chunk_tags]
-        tags, field_name = self.convert_bioul_to_segmental(chunk_tags)
-        instance_fields[field_name] = SequenceLabelField(tags, sentence, field_name)
+        tags, namespace = self.convert_bioul_to_segmental(chunk_tags)
+        instance_fields["tags"] = SequenceLabelField(tags, sentence, namespace)
```
Comment (brendan-ai2) [inline allennlp/data/dataset_readers/segmental_conll2000.py:168]: Why did this change?

File: allennlp/models/language_model.py [hunk 2]
```diff
@@ -294,25 +297,29 @@ def forward(self,  # type: ignore
...
                         'loss': average_loss,
-                        'forward_loss': forward_loss / num_targets.float(),
-                        'backward_loss': (backward_loss / num_targets.float()
+                        'forward_loss': forward_loss / num_targets,
+                        'backward_loss': (backward_loss / num_targets
                                           if backward_loss is not None else None),
-                        'batch_weight': num_targets.float()
+                        'batch_weight': num_targets
                 })
+
```
Comment (brendan-ai2) [inline allennlp/models/language_model.py:322]: Checking my understanding, I think the above lines are just refactoring, no functional changes other than adding the metric. Is that right?

File: allennlp/models/segmental_language_model.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+from allennlp.modules.seq2seq_encoders import Seq2SeqEncoder
+from allennlp.nn.util import get_text_field_mask
+from allennlp.nn import InitializerApplicator
+from allennlp.training.metrics.perplexity import Perplexity
+
+
+@Model.register('segmental_language_model')
+class SegmentalLanguageModel(LanguageModel):
+    """
+    The ``LanguageModel`` applies a "contextualizing"
```
Comment (brendan-ai2) [inline allennlp/models/segmental_language_model.py:24]: Nit: I think this is a duplicate of the `LanguageModel` comment, I'd delete and say "See ``LanguageModel`` for details." Bonus points for describing what makes the SegmentalLanguageModel different, but that can probably wait for after ACL. ;)

File: allennlp/models/segmental_language_model.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+                 sparse_embeddings: bool = False,
+                 bidirectional: bool = True,
+                 initializer: InitializerApplicator = None) -> None:
+        super().__init__(vocab=vocab,
+                         text_field_embedder=text_field_embedder,
+                         contextualizer=contextualizer,
+                         dropout=dropout,
+                         num_samples=num_samples,
+                         sparse_embeddings=sparse_embeddings,
+                         bidirectional=bidirectional,
```
Comment (brendan-ai2) [inline allennlp/models/segmental_language_model.py:84]: I think this is always true for your use-case, right?

File: allennlp/models/segmental_language_model.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+                         initializer=initializer)
+        self._forward_segmental_contextualizer = forward_segmental_contextualizer
+        self._backward_segmental_contextualizer = backward_segmental_contextualizer
+
+        if num_samples is not None:
```
Comment (brendan-ai2) [inline allennlp/models/segmental_language_model.py:89]: I'd add a comment here for why you need a different softmax than the regular LM. I think it's due to the forward dimension size?

File: allennlp/models/segmental_language_model.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+        Computes the averaged forward (and backward, if language model is bidirectional)
+        LM loss from the batch.
+
+        By convention, the input dict is required to have at least a ``"tokens"``
+        entry that's the output of a ``SingleIdTokenIndexer``, which is used
+        to compute the language model targets.
+
+        Parameters
+        ----------
+        tokens: ``torch.Tensor``, required.
```
Comment (brendan-ai2) [inline allennlp/models/segmental_language_model.py:133]: Could you update this comment? It's hard to follow with the old ones.

File: allennlp/models/segmental_language_model.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+        # Secondly, the segmental embeddings need to be scattered, so each
+        # position gets its own segmental information.
+        seg_embeddings_scattered, _ = self._get_gathered_embeddings(
+            embeddings=seg_embeddings_with_dropout,
+            indices=mapping)
+        return seg_embeddings_scattered
+
+    @staticmethod
+    def _get_gathered_embeddings(embeddings: torch.Tensor,
```
Comment (brendan-ai2) [inline allennlp/models/segmental_language_model.py:285]: Could you add some more comments for these last two methods? What they do at a high level, basically. I was having a bit of trouble following.

File: allennlp/modules/seq2seq_encoders/bidirectional_language_model_transformer.py [hunk 8]
```diff
@@ -259,16 +265,34 @@ def forward(self, token_embeddings: torch.Tensor, mask: torch.Tensor) -> torch.T
...
+                return to_return
+            return torch.cat([forward_output, backward_output], -1)
+        elif self.go_forward and not self.go_backward:
+            if self._return_all_layers:
+                to_return = []
+                for forward in forward_output:
+                    to_return.append(forward)
+                return to_return
+            return forward_output
+        elif self.go_backward and not self.go_forward:
```
Comment (brendan-ai2) [inline allennlp/modules/seq2seq_encoders/bidirectional_language_model_transformer.py:288]: Nit: I think the redundancy in this block can be reduced by doing something like:
```
            if self._return_all_layers:
                to_return = []
                for forward, backward in zip(forward_output, backward_output):
                    layers = []
                    if self.go_forward:
                        layers.append(forward)
                    if self.go_backward:
                        layers.append(backward)
                    to_return.append(torch.cat(layers, -1))
                return to_return
            layers = []
            if self.go_forward:
                layers.append(forward_output)
            if self.go_backward:
                layers.append(backward_output)
            return torch.cat(layers, -1)
```

File: training_config/segmental_language_model.jsonnet [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+      // words not used in the sampled softmax would be decayed to zero with the
+      // standard "adam" optimizer.
+      "type": "dense_sparse_adam"
+    },
+    // TODO(brendanr): Needed with transformer too?
+    // "grad_norm": 10.0,
+    "learning_rate_scheduler": {
+      "type": "noam",
+      // See https://github.com/allenai/calypso/blob/master/calypso/train.py#L401
+      "model_size": 512,
```
Comment (brendan-ai2) [inline training_config/segmental_language_model.jsonnet:155]: This probably needs tweaking...

File: training_config/segmental_language_model.jsonnet [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        "num_layers": 6,
+        "dropout": 0.1,
+        "input_dropout": 0.1
+    },
+    "forward_segmental_contextualizer": {
+      "type": "bidirectional_language_model_transformer",
+      "input_dim": 512,
+      "hidden_dim": 2048,
+      "input_dropout": 0.1,
+      "num_layers": 2,
```
Comment (brendan-ai2) [inline training_config/segmental_language_model.jsonnet:116]: This is kinda shallow for a transformer, no? I'm curious about whether you tried different things here.

File: allennlp/models/segmental_language_model.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+        )
+
+        return_dict = {'lm_embeddings': contextual_embeddings,
+                       'noncontextual_token_embeddings': embeddings,
+                       'mask': mask
+                       }
+
+        token_ids = tokens.get("tokens")
+        if token_ids is None:
+            return return_dict
```
Comment (brendan-ai2) [inline allennlp/models/segmental_language_model.py:175]: Wait, so there isn't any segmental info in the case where the LM is being used as an embedder? I think I need to plead for one more overview of the entire system tomorrow!

File: allennlp/models/segmental_language_model.py [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+                                                    sparse=sparse_embeddings)
+        else:
+            self._softmax_loss = _SoftmaxLoss(num_words=vocab.get_vocab_size(),
+                                              embedding_dim=softmax_projection_dim)
+
+        self.num_classes = self.vocab.get_vocab_size(label_namespace)
+        self.label_feature_embedding = Embedding(self.num_classes, label_feature_dim)
+
+        self._forward_dim = contextualizer.get_output_dim() // 2 + \
+                            forward_segmental_contextualizer.get_output_dim() // 2 + \
```
Comment (brendan-ai2) [inline allennlp/models/segmental_language_model.py:102]: Why `// 2` here?

**Review by brendan-ai2 — COMMENTED** [review]
Hey! Thanks for sending this!! So here are some long overdue comments. A decent number are to help clarify my understanding. Most of the "cleanup" requests aren't really urgent with the possible exception of the requests for more comments. Without those it's pretty hard to follow.

Looking forward to those experiments tomorrow!

File: allennlp/models/language_model.py (hunk 2 above)
Comment (swabhs) [reply allennlp/models/language_model.py:322]: Exactly, except for the inclusion of the `Perplexity` metric.
